File: problem_5/Solution.java
/*
  2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.
  What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?
*/

import java.util.*;

class Solution {
  public static void main(String[] args) {
    assert( numberToPrimes(20).equals(new ArrayList<Integer>() {{ add(2);add(2);add(5); }} ) );
    assert( numberToPrimes(19).equals(new ArrayList<Integer>() {{ add(19); }} ) );

    ArrayList<Integer> a1 = new ArrayList<Integer>() {{ add(2);add(2);add(3); }};
    ArrayList<Integer> a2 = new ArrayList<Integer>() {{ add(2);add(5);add(4); }};
    assert(mergeArrays(a1,a2).equals(new ArrayList<Integer>() {{ add(2);add(2);add(3);add(5);add(4); }} ) );

    System.out.println(getSmallestNumberDivisibleBySeries());
  }

  public static Integer getSmallestNumberDivisibleBySeries() {
    ArrayList<Integer> result = new ArrayList<Integer>();
    ArrayList<Integer> multipliers = new ArrayList<Integer>();
    for(int i = 20; i > 1; i--) {
      multipliers = numberToPrimes(i);
      result = mergeArrays(multipliers, result);
    }
    Integer resultNumber = 1;
    for(Integer m : result) {
      resultNumber *= m;
    }
    return resultNumber;
  }

  public static ArrayList<Integer> mergeArrays(ArrayList<Integer> a, ArrayList<Integer> b) {
    ArrayList<Integer> result = new ArrayList<Integer>();
    Integer temp = 0;
    int index = 0;
    Iterator<Integer> itr = a.iterator();
    while(itr.hasNext()) {
      temp = itr.next();
      itr.remove();
      index = b.indexOf(temp);
      if(index >= 0) {
        b.remove(index);
      }
      result.add(temp);
    }
    result.addAll(b);
    return result;
  }

  public static ArrayList<Integer> numberToPrimes(Integer number) {
    ArrayList<Integer> result = new ArrayList<Integer>();
    Integer a = 2;
    Integer max = number;
    while(a<=max) {
      if(number % a == 0) {
        while(number % a == 0) {
          result.add(a);
          number /= a;
        }
      }
      a++;
    }
    return result;
  }
}

File: problem_5/Solution.scala
import scala.collection.mutable.ListBuffer;

object Solution {
  def main(args: Array[String]) {
    println(numberToPrimes(20))
    println(mergeArrays(ListBuffer(2,2,3), ListBuffer(2,2)))
    println(getSmallestNumberDivisibleBySeries())
  }

  def getSmallestNumberDivisibleBySeries() = {
    var result = new ListBuffer[Int]
    var multipliers = new ListBuffer[Int]

    for (i <- 0 until 20) {
      multipliers = numberToPrimes(i)
      result = mergeArrays(multipliers, result)
    }
    var resultNumber = 1
    for (ii <- result) {
      resultNumber *= ii
    }
    resultNumber
  }

  def mergeArrays(a:ListBuffer[Int], b:ListBuffer[Int]) = {
    var result = new ListBuffer[Int]
    var index = 0
    var aa = a
    var bb = b
    for (el <- aa) {
      if(bb.indexOf(el) >= 0) {
        bb.remove(bb.indexOf(el))
      }
      result += el
      aa.remove(aa.indexOf(el))
    }
    result ++ bb
  }

  def numberToPrimes(number:Int) = {
    var result = new ListBuffer[Int]
    var nnumber = number
    var i = 2
    while(i<=number) {
      if(nnumber % i == 0) {
        while(nnumber % i ==0) {
          nnumber /= i
          result += i
        }
      }
      i += 1
    }
    result
  }
}
